SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorPackageAssembly.cpp

```diff
@@ -0,0 +1,242 @@
+			});
+			break;
+		case EPackageAssemblyStep::BUILD_CLIENT:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FApp::GetProjectName(), Win64, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::BUILD_SIMULATED_PLAYERS:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FString::Printf(TEXT("%sSimulatedPlayer"), FApp::GetProjectName()), Linux, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::UPLOAD_ASSEMBLY:
+		{
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->AssemblyDetailsPtr->Upload(*this);
+			});
+		}
+		break;
+		}
+	}
+	return HasMoreSteps;
+}
+
+void FSpatialGDKPackageAssembly::OnTaskCompleted(int32 TaskResult)
+{
+	if (TaskResult == 0)
+	{
+		if (!NextStep())
+		{
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				FString NotificationMessage = FString::Printf(TEXT("Assembly successfully uploaded to project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+				this->ShowTaskEndedNotification(NotificationMessage, SNotificationItem::CS_Success);
+				OnSuccess.ExecuteIfBound();
+			});
+		}
+	}
+	else
+	{
+		FString NotificationMessage = FString::Printf(TEXT("Failed assembly upload to project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+		AsyncTask(ENamedThreads::GameThread, [this]() { this->ShowTaskEndedNotification(TEXT("Assembly Failed"), SNotificationItem::CS_Fail); });
+		Steps.Empty();
+	}
+}
+
+void FSpatialGDKPackageAssembly::OnTaskOutput(FString Message)
+{
+	UE_LOG(LogSpatialGDKEditorPackageAssembly, Display, TEXT("%s"), *Message);
+}
+
+void FSpatialGDKPackageAssembly::OnTaskCanceled()
+{
+	Steps.Empty();
+	FString NotificationMessage = FString::Printf(TEXT("Cancelled assembly upload to project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+	AsyncTask(ENamedThreads::GameThread, [this, NotificationMessage]()
+	{
+		this->ShowTaskEndedNotification(NotificationMessage, SNotificationItem::CS_Fail);
+		this->AssemblyDetailsPtr.Reset();
+	});
+}
+
+FSpatialGDKPackageAssembly::AssemblyDetails::AssemblyDetails(const FString& Name, const FString& Config, bool bInForce)
+	: AssemblyName(Name)
+	, Configuration(Config)
+	, bForce(bInForce)
+{
+}
+
+void FSpatialGDKPackageAssembly::AssemblyDetails::Upload(FSpatialGDKPackageAssembly& PackageAssembly)
+{
+	PackageAssembly.UploadAssembly(AssemblyName, bForce);
+}
+
+void FSpatialGDKPackageAssembly::HandleCancelButtonClicked()
+{
+	if (PackageAssemblyTask.IsValid())
+	{
+		PackageAssemblyTask->Cancel(true);
+	}
+}
+
+void FSpatialGDKPackageAssembly::ShowTaskStartedNotification(const FString& NotificationText)
+{
+	FNotificationInfo Info(FText::AsCultureInvariant(NotificationText));
+	Info.ButtonDetails.Add(
+		FNotificationButtonInfo(
+			LOCTEXT("PackageAssemblyTaskCancel", "Cancel"),
```

Comment:
Ahhh smart. I should remember to add localized text too. I guess PackageAssemblyTaskCancel is the localized text key. Cool stuff.